File: backend/app/utils/time_utils.py
"""
Common time utility functions for consistent timestamp handling
Used across currency rates and crypto prices
"""
from datetime import datetime, timezone
from typing import Optional, Dict, Any


def format_timestamp(timestamp: Optional[datetime]) -> str:
    """Format timestamp for display (e.g., '2025-10-24 19:22:09 UTC')"""
    if not timestamp:
        return "Never updated"
    
    try:
        # Ensure timestamp is timezone-aware
        if timestamp.tzinfo is None:
            timestamp = timestamp.replace(tzinfo=timezone.utc)
        
        return timestamp.strftime("%Y-%m-%d %H:%M:%S UTC")
    except Exception:
        return "Invalid timestamp"


def get_iso_timestamp(timestamp: Optional[datetime] = None) -> str:
    """Get ISO format timestamp for API responses"""
    if timestamp:
        # Ensure timestamp is timezone-aware
        if timestamp.tzinfo is None:
            timestamp = timestamp.replace(tzinfo=timezone.utc)
        return timestamp.isoformat()
    else:
        return datetime.now(timezone.utc).isoformat()


def get_current_timestamp() -> datetime:
    """Get current timestamp in UTC"""
    return datetime.now(timezone.utc)


def is_valid_timestamp(timestamp: Any) -> bool:
    """Check if timestamp is valid"""
    if not timestamp:
        return False
    
    try:
        if isinstance(timestamp, datetime):
            return True
        elif isinstance(timestamp, str):
            datetime.fromisoformat(timestamp.replace('Z', '+00:00'))
            return True
        return False
    except (ValueError, TypeError):
        return False


def get_timestamp_display_info(timestamp: Optional[datetime]) -> Dict[str, str]:
    """Get comprehensive timestamp display information"""
    if not timestamp:
        return {
            "formatted": "Never updated",
            "iso": "",
            "relative": "Unknown"
        }
    
    return {
        "formatted": format_timestamp(timestamp),
        "iso": get_iso_timestamp(timestamp),
        "relative": get_relative_time(timestamp)
    }


def get_relative_time(timestamp: datetime) -> str:
    """Get relative time string (e.g., '2h ago', '30s ago')"""
    try:
        now = datetime.now(timezone.utc)
        
        # Ensure timestamp is timezone-aware
        if timestamp.tzinfo is None:
            timestamp = timestamp.replace(tzinfo=timezone.utc)
        
        diff = now - timestamp
        total_seconds = int(diff.total_seconds())
        
        if total_seconds < 60:
            return f"{total_seconds}s ago"
        elif total_seconds < 3600:
            minutes = total_seconds // 60
            return f"{minutes}m ago"
        elif total_seconds < 86400:
            hours = total_seconds // 3600
            return f"{hours}h ago"
        else:
            days = total_seconds // 86400
            return f"{days}d ago"
    except Exception:
        return "Invalid time"


def get_data_freshness(timestamp: Optional[datetime]) -> str:
    """Get data freshness status based on timestamp age"""
    if not timestamp:
        return "stale"
    
    try:
        now = datetime.now(timezone.utc)
        
        # Ensure timestamp is timezone-aware
        if timestamp.tzinfo is None:
            timestamp = timestamp.replace(tzinfo=timezone.utc)
        
        diff = now - timestamp
        diff_minutes = int(diff.total_seconds() / 60)
        
        if diff_minutes < 1:
            return "fresh"
        elif diff_minutes < 5:
            return "recent"
        else:
            return "stale"
    except Exception:
        return "stale"

File: frontend/src/lib/timeUtils.ts
/**
 * Common time utility functions for consistent timestamp handling
 * Used across currency rates and crypto prices
 */

export type DataFreshness = 'fresh' | 'recent' | 'stale';

export interface TimestampDisplay {
  formatted: string;
  relative: string;
  freshness: DataFreshness;
  iso: string;
}

/**
 * Get relative time string (e.g., "2h ago", "30s ago")
 */
export const getRelativeTime = (timestamp: string): string => {
  if (!timestamp) return 'Unknown';
  
  try {
    const now = new Date();
    const time = new Date(timestamp);
    const diffMs = now.getTime() - time.getTime();
    const diffMinutes = Math.floor(diffMs / (1000 * 60));
    const diffSeconds = Math.floor(diffMs / 1000);
    
    if (diffSeconds < 60) {
      return `${diffSeconds}s ago`;
    } else if (diffMinutes < 60) {
      return `${diffMinutes}m ago`;
    } else if (diffMinutes < 1440) {
      const hours = Math.floor(diffMinutes / 60);
      return `${hours}h ago`;
    } else {
      const days = Math.floor(diffMinutes / 1440);
      return `${days}d ago`;
    }
  } catch (error) {
    return 'Invalid time';
  }
};

/**
 * Get data freshness status based on timestamp age
 */
export const getDataFreshness = (timestamp: string): DataFreshness => {
  if (!timestamp) return 'stale';
  
  try {
    const now = new Date();
    const time = new Date(timestamp);
    const diffMs = now.getTime() - time.getTime();
    const diffMinutes = Math.floor(diffMs / (1000 * 60));
    
    if (diffMinutes < 1) return 'fresh';
    if (diffMinutes < 5) return 'recent';
    return 'stale';
  } catch (error) {
    return 'stale';
  }
};

/**
 * Format timestamp for display (e.g., "2025-10-24 19:22:09 UTC")
 */
export const formatTimestamp = (timestamp: string): string => {
  if (!timestamp) return 'Never updated';
  
  try {
    const date = new Date(timestamp);
    return date.toLocaleString('en-US', {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit',
      timeZone: 'UTC',
      timeZoneName: 'short'
    });
  } catch (error) {
    return 'Invalid timestamp';
  }
};

/**
 * Get comprehensive timestamp display information
 */
export const getTimestampDisplay = (timestamp: string): TimestampDisplay => {
  return {
    formatted: formatTimestamp(timestamp),
    relative: getRelativeTime(timestamp),
    freshness: getDataFreshness(timestamp),
    iso: timestamp
  };
};

/**
 * Get freshness color class for UI
 */
export const getFreshnessColorClass = (freshness: DataFreshness): string => {
  switch (freshness) {
    case 'fresh':
      return 'bg-green-500';
    case 'recent':
      return 'bg-yellow-500';
    case 'stale':
      return 'bg-red-500';
    default:
      return 'bg-gray-500';
  }
};

/**
 * Check if timestamp is valid
 */
export const isValidTimestamp = (timestamp: string): boolean => {
  if (!timestamp) return false;
  try {
    const date = new Date(timestamp);
    return !isNaN(date.getTime());
  } catch (error) {
    return false;
  }
};

/**
 * Get current timestamp in ISO format
 */
export const getCurrentTimestamp = (): string => {
  return new Date().toISOString();
};
